### PR TITLE
feat: zeta supply checks

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

- Adds helper functions and a thread to the zetaclient to perform the zeta supply checks.

Closes: <#1352>

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: app/setup_handlers.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ func SetupHandlers(app *App) {
 		}
 		vm[observertypes.ModuleName] = vm[observertypes.ModuleName] - 1
 		return app.mm.RunMigrations(ctx, app.configurator, vm)
+
```
Comment (lumtis) [inline app/setup_handlers.go:22]:
```suggestion
```

File: cmd/zetaclientd/start.go [hunk 2]
```diff
@@ -240,6 +241,14 @@ func start(_ *cobra.Command, _ []string) error {
 	mo1 := mc.NewCoreObserver(zetaBridge, signerMap, chainClientMap, metrics, tss, masterLogger, cfg, telemetryServer)
 	mo1.MonitorCore()
 
+	zetaSupplyChecker, errStartingzetaSupplyChecker := mc.NewZetaSupplyChecker(cfg, zetaBridge, masterLogger)
```
Comment (lumtis) [inline cmd/zetaclientd/start.go:244]:
```suggestion
	zetaSupplyChecker, errStartingZetaSupplyChecker := mc.NewZetaSupplyChecker(cfg, zetaBridge, masterLogger)
```

Comment (lumtis) [inline cmd/zetaclientd/start.go:244]:
```suggestion
	zetaSupplyChecker, err := mc.NewZetaSupplyChecker(cfg, zetaBridge, masterLogger)
```
Why not just `err`?

File: zetaclient/zeta_supply_checker_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			zetaInTransit:            MustNewIntFromString("1000000000000000000"),
+			externalChainTotalSupply: MustNewIntFromString("9000000000000000000"),
+			genesisAmounts:           MustNewIntFromString("1000000000000000000"),
+			zetaTokenSupplyOnNode:    MustNewIntFromString("1000000000000000000"),
+			ethLockedAmount:          MustNewIntFromString("10000000000000000000"),
+			validate: func(t assert.TestingT, b bool, i ...interface{}) bool {
+				return assert.True(t, b, i...)
+			},
+		},
+		// Todo add more scenarios
```
Comment (lumtis) [inline zetaclient/zeta_supply_checker_test.go:43]: We should track issue for TODOs

**Review by lumtis — COMMENTED** [review]
As far as I can tell `CctxByStatus` would be called every ticks?
It would be a very intensive operation iterating the cctxs in this query.

If we just need this query for aborted CCTXs we should think about a new indexing like we do for pending cctxs, or storing the result we want for the aborted cctxs directly onchain.

Another solution is to use a backend indexer for the data but in this case we should no longer consider observers posting invariant failure to the protocol because they should only rely on blockchain data in order to do this.

File: zetaclient/zeta_supply_checker.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+			return err
+		}
+		totalSupply, err := zetatokenNonEth.TotalSupply(nil)
+		if err != nil {
+			return err
+		}
+		totalSupplyInt, ok := sdkmath.NewIntFromString(totalSupply.String())
+		if !ok {
+			zs.logger.Error().Msgf("error parsing total supply for chain %d", chain.ChainId)
+			continue
```
Comment (ws4charlie) [inline zetaclient/zeta_supply_checker.go:113]: should we `break` like above when conversion fails?

File: zetaclient/zeta_supply_checker.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	logger.Info().Msgf("zeta supply check passed, lhs : %s , rhs : %s", lhs.String(), rhs.String())
+	logger.Info().Msgf("--------------------------------------------------------------------------------")
+	return true
+}
+
+func (zs *ZetaSupplyChecker) AbortedTxAmount() sdkmath.Int {
+	cctxList, err := zs.zetaClient.GetCctxByStatus(types.CctxStatus_Aborted)
+	if err != nil {
+		panic(err)
```
Comment (ws4charlie) [inline zetaclient/zeta_supply_checker.go:174]: maybe allowing `AbortedTxAmount` throwing an error and avoid panic is better.

File: zetaclient/zeta_supply_checker.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	amount := sdkmath.ZeroUint()
+	for _, cctx := range cctxList {
+		amount = amount.Add(cctx.GetCurrentOutTxParam().Amount)
+	}
+	amountInt, ok := sdkmath.NewIntFromString(amount.String())
+	if !ok {
+		panic("error parsing amount")
```
Comment (ws4charlie) [inline zetaclient/zeta_supply_checker.go:182]: avoid panic here too?

File: zetaclient/zeta_supply_checker.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+			}
+		}
+
+		trackers, err := zs.zetaClient.GetAllOutTxTrackerByChain(chain, Ascending)
+		if err != nil {
+			continue
+		}
+		for _, tracker := range trackers {
+			zs.logger.Info().Msgf("tracker exists for nonce: %d , removing from supply checks", tracker.Nonce)
+			delete(nonceToCctxMap, tracker.Nonce)
```
Comment (ws4charlie) [inline zetaclient/zeta_supply_checker.go:221]: the purpose here is to exclude pending cctcs who have a tracker?

File: zetaclient/zeta_supply_checker.go (hunk 4 above)
Comment (kingpinXD) [reply zetaclient/zeta_supply_checker.go:113]: I guess we can continue adding up the supplies to see if the final check passes

File: zetaclient/zeta_supply_checker.go (hunk 7 above)
Comment (kingpinXD) [reply zetaclient/zeta_supply_checker.go:221]: Yes

**Comment by kingpinXD** [comment]
> As far as I can tell `CctxByStatus` would be called every ticks? It would be a very intensive operation iterating the cctxs in this query.
> 
> If we just need this query for aborted CCTXs we should think about a new indexing like we do for pending cctxs, or storing the result we want for the aborted cctxs directly onchain.
> 
> Another solution is to use a backend indexer for the data but in this case we should no longer consider observers posting invariant failure to the protocol because they should only rely on blockchain data in order to do this.

Yeah, I thought of this; since we only really need the amount, storing that separately might be a good idea. But we can let this run for sometime on athens3 to see how well it holds and implement a service to keep track of this amount only.
The GetCctxByStatus is just a good-to-have query, and I agree that it should not be used for this purpose.

File: zetaclient/zeta_supply_checker.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	sdkmath "cosmossdk.io/math"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/rs/zerolog"
+	"github.com/zeta-chain/zetacore/common"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/zetaclient/config"
+)
+
+type ZetaSupplyChecker struct {
```
Comment (lumtis) [inline zetaclient/zeta_supply_checker.go:17]: Can we define an interface for `ZetaSupplyChecker`?
We're going to use interfaces for ZetaClient unit tests which will be one of the next milestone for the ZetaClient

**Review by lumtis — COMMENTED** [review]
>Yeah, I thought of this; since we only really need the amount, storing that separately might be a good idea. But we can let this run for sometime on athens3 to see how well it holds and implement a service to keep track of this amount only.


I think we can keep it as it for experimentation but we should have a more optimized solution for poroduction.

> The GetCctxByStatus is just a good-to-have query, and I agree that it should not be used for this purpose.

I think we should avoid having queries reading extensively in the store, here going through all cctxs. We should use paginated read in the store which is here not possible since we read and filter the object. I think the best is to have an additional indexation for the specific cctx status that we need.

**Review by lumtis — APPROVED** [review]

**Review by ws4charlie — APPROVED** [review]
looks good

**Pull request merged**